src/goto-programs/remove_instanceof.cpp

```diff
@@ -42,160 +42,129 @@ class remove_instanceoft
-/// instruction `this_inst` it belongs to.
-/// \return Side-effect on `expr` replacing it with an explicit clsid test
-void remove_instanceoft::lower_instanceof(
+/// \param expr: Expression to lower (the code or the guard of an instruction)
+/// \param goto_program: program the expression belongs to
+/// \param this_inst: instruction the expression is found at
+/// \return number of instanceof expressions that have been replaced
+int remove_instanceoft::lower_instanceof(
   exprt &expr,
   goto_programt &goto_program,
-  goto_programt::targett this_inst,
-  instanceof_instt &inst_switch)
+  goto_programt::targett this_inst)
 {
-  if(expr.id()==ID_java_instanceof)
+  if(expr.id()!=ID_java_instanceof)
   {
-    const exprt &check_ptr=expr.op0();
-    assert(check_ptr.type().id()==ID_pointer);
-    const exprt &target_arg=expr.op1();
-    assert(target_arg.id()==ID_type);
-    const typet &target_type=target_arg.type();
-
-    // Find all types we know about that satisfy the given requirement:
-    assert(target_type.id()==ID_symbol);
-    const irep_idt &target_name=
-      to_symbol_type(target_type).get_identifier();
-    std::vector<irep_idt> children=
-      class_hierarchy.get_children_trans(target_name);
-    children.push_back(target_name);
-
-    assert(!children.empty() && "Unable to execute instanceof");
-
-    // Insert an instruction before this one that assigns the clsid we're
-    // checking against to a temporary, as GOTO program if-expressions should
-    // not contain derefs.
-
-    symbol_typet jlo("java::java.lang.Object");
-    exprt object_clsid=get_class_identifier_field(check_ptr, jlo, ns);
-
-    symbolt &newsym=
-      get_fresh_aux_symbol(
-        object_clsid.type(),
-        "instanceof_tmp",
-        "instanceof_tmp",
-        source_locationt(),
-        ID_java,
-        symbol_table);
-
-    auto newinst=goto_program.insert_after(this_inst);
-    newinst->make_assignment();
-    newinst->code=code_assignt(newsym.symbol_expr(), object_clsid);
-    newinst->source_location=this_inst->source_location;
-    newinst->function=this_inst->function;
-
-    // Insert the check instruction after the existing one.
-    // This will briefly be ill-formed (use before def of
-    // instanceof_tmp) but the two will subsequently switch
-    // places in lower_instanceof(goto_programt &) below.
-    inst_switch.push_back(make_pair(this_inst, newinst));
-    // Replace the instanceof construct with a big-or.
-    exprt::operandst or_ops;
-    for(const auto &clsname : children)
-    {
-      constant_exprt clsexpr(clsname, string_typet());
-      equal_exprt test(newsym.symbol_expr(), clsexpr);
-      or_ops.push_back(test);
-    }
-    expr=disjunction(or_ops);
+    int replacements=0;
+    Forall_operands(it, expr)
+      replacements+=lower_instanceof(*it, goto_program, this_inst);
+    return replacements;
   }
-  else
+
+  const exprt &check_ptr=expr.op0();
```

Comment:
Add an INVARIANT that says that 'expr' has the right number of operands (or add a java_instanceof_exprt, with corresponding to_java_instanceof_expr() function).